src/water_balance.jl

```diff
@@ -0,0 +1,97 @@
+struct WaterBalanceVertical{T <: AbstractFloat} # Terms in [mm]
+    storage_total::Vector{T}
+    storage_total_prev::Vector{T}
+    inflow_total::Vector{T}
+    outflow_total::Vector{T}
+    balance_error::Vector{T}
+    relative_error::Vector{T}
+end
+
+function WaterBalanceVertical(n::Integer, float_type::Type{T}) where {T <: AbstractFloat}
+    WaterBalanceVertical(ntuple(_ -> fill(float_type(Nan), n), 6)...)
+end
+
+function compute_water_balance!(model, water_balance::WaterBalanceVertical)::Nothing
+    (;
+        storage_total,
+        storage_total_prev,
+        inflow_total,
+        outflow_total,
+        balance_error,
+        relative_error,
+    ) = water_balance
+    (; vertical, lateral) = model
+
+    (; soil, runoff) = vertical
+    (; satwaterdepth, ustoredepth, actevap) = soil.variables
+    (; runoff_land, net_runoff_river) = runoff.variables
+
+    (; subsurface) = lateral
+    (; ssf, ssfin) = subsurface.variables
+    (; flow_length, flow_width) = subsurface.parameters
+
+    storage_total .= satwaterdepth
+    storage_total .+= ustoredepth
+
+    inflow_total .= vertical.atmospheric_forcing.precipitation
+    inflow_total .+= runoff.variables.net_runoff_river
```

Comment:
`net_runoff_river` is not an inflow (it is an outflow). `allocation` refers to `model.land.allocation`.
```suggestion
    inflow_total .+= get_irrigation_allocated(allocation)
```